src/base/sparsity_pattern.C

```diff
@@ -448,123 +448,164 @@ void Build::parallel_sync ()
+  while (it != nonlocal_pattern.end())
+  {
+    const auto dof_id = it->first;
+    auto & row = it->second;
 
-      const std::size_t n_rows = pushed_row_ids_to_me.size();
-      for (std::size_t i=0; i != n_rows; ++i)
-        {
-          const dof_id_type r = pushed_row_ids_to_me[i];
-          const dof_id_type my_r = r - local_first_dof;
+    processor_id_type proc_id = 0;
+    while (dof_id >= dof_map.end_dof(proc_id))
+      proc_id++;
 
-          std::vector<dof_id_type> & their_row = pushed_rows_to_me[i];
+    will_send_to[proc_id] = true;
+    num_sends++;
 
-          if (need_full_sparsity_pattern)
-            {
-              SparsityPattern::Row & my_row =
-                sparsity_pattern[my_r];
+    // rhs [] on purpose
+    auto & proc_data = data_to_send[proc_id];
 
-              // They wouldn't have sent an empty row
-              libmesh_assert(!their_row.empty());
+    proc_data.first.push_back(dof_id);
 
-              // We can end up with an empty row on a dof that touches our
-              // inactive elements but not our active ones
-              if (my_row.empty())
-                {
-                  my_row.assign (their_row.begin(),
-                                 their_row.end());
-                }
-              else
-                {
-                  my_row.insert (my_row.end(),
-                                 their_row.begin(),
-                                 their_row.end());
+    // Note this invalidates the data in nonlocal_pattern
+    proc_data.second.emplace_back(std::move(row));
 
-                  // We cannot use SparsityPattern::sort_row() here because it expects
-                  // the [begin,middle) [middle,end) to be non-overlapping.  This is not
-                  // necessarily the case here, so use std::sort()
-                  std::sort (my_row.begin(), my_row.end());
+    // Might as well remove it since it's invalidated anyway
+    nonlocal_pattern.erase(it++);
+  }
 
-                  my_row.erase(std::unique (my_row.begin(), my_row.end()), my_row.end());
-                }
+  // Tell everyone about where everyone will send to
+  comm.alltoall(will_send_to);
 
-              // fix the number of on and off-processor nonzeros in this row
-              n_nz[my_r] = n_oz[my_r] = 0;
+  // will_send_to now represents who we'll receive from
+  // give it a good name
+  auto & will_receive_from = will_send_to;
 
-              for (std::size_t j=0; j<my_row.size(); j++)
-                if ((my_row[j] < local_first_dof) || (my_row[j] >= local_end_dof))
-                  n_oz[my_r]++;
-                else
-                  n_nz[my_r]++;
-            }
-          else
-            {
-              for (std::size_t j=0; j<their_row.size(); j++)
-                if ((their_row[j] < local_first_dof) || (their_row[j] >= local_end_dof))
-                  n_oz[my_r]++;
-                else
-                  n_nz[my_r]++;
+  std::vector<Parallel::Request> dof_sends(num_sends);
```

Comment:
Should we check that num_sends is the same as std::accumulate of will_send_to?  I think this should be a symmetric communication but I wouldn't want to swear to it.

Reply:
(by which I mean "of will_receive_from", after the alltoall)

Reply:
No, it shouldn't be a symmetric communication, but that doesn't matter; your code looks like it should be doing the correct number of receives.